WebKit needs to continue "wrongly" re-exporting libobjc.dylib for compatibility

emw-apple opened the pull request:
#### c11c33ec0e67be5c6bf3da75dd713a0c0b61c083
<pre>
WebKit needs to continue &quot;wrongly&quot; re-exporting libobjc.dylib for compatibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=264050">https://bugs.webkit.org/show_bug.cgi?id=264050</a>
<a href="https://rdar.apple.com/117360317">rdar://117360317</a>

Reviewed by Alexey Proskuryakov.

libobjc is re-exported by WebCore, and WebKit on macOS stopped
re-exporting WebCore in <a href="https://bugs.webkit.org/show_bug.cgi?id=259352.">https://bugs.webkit.org/show_bug.cgi?id=259352.</a>
Some clients linked against WebKit and not libobjc, and recorded the
wrong two-level namespace for objc symbols.

Make WebKit re-export libobjc to fix these applications.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: <a href="https://commits.webkit.org/270085@main">https://commits.webkit.org/270085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fc68f8fc22ba361703c3445f3272c981eace629

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26633 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4699 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/488 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24750 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/2122 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27218 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1853 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28295 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22346 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26087 "Found 3 new API test failures: /WebKitGTK/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WebKitGTK/TestLoaderClient:/webkit/WebKitURIRequest/http-headers, /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/handle-corrupted-local-storage (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1787 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3097 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5870 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2244 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -221,6 +221,8 @@ WK_NO_STATIC_INITIALIZERS_Production_NO_ = $(WK_ERROR_WHEN_LINKING_WITH_STATIC_I
 
 REEXPORTED_FRAMEWORK_NAMES = WebKitLegacy;
+// Needed for bincompat (rdar://117360317)
+REEXPORTED_LIBRARY_NAMES[sdk=macosx*] = objc;
```
This won't overwrite the above line, and stop re-exporting WebKitLegacy?

emw-apple replied:
It does not, and after installing a root with this change, I see both libraries as re-exports:
```
> dyld_info -dependents /System/Library/Frameworks/WebKit.framework/Versions/A/WebKit
/System/Library/Frameworks/WebKit.framework/Versions/A/WebKit [arm64e]:
    -dependents:
        attributes     load path
        re-export      /usr/lib/libobjc.A.dylib
        re-export      /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebKitLegacy.framework/Versions/A/WebKitLegacy
...
```

aproskuryakov replied:
My bad, didn't notice that these were distinct settings.

emw-apple committed:
WebKit needs to continue "wrongly" re-exporting libobjc.dylib for compatibility
https://bugs.webkit.org/show_bug.cgi?id=264050
rdar://117360317

Reviewed by Alexey Proskuryakov.

libobjc is re-exported by WebCore, and WebKit on macOS stopped
re-exporting WebCore in https://bugs.webkit.org/show_bug.cgi?id=259352.
Some clients linked against WebKit and not libobjc, and recorded the
wrong two-level namespace for objc symbols.

Make WebKit re-export libobjc to fix these applications.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: https://commits.webkit.org/270085@main
